Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs

```diff
@@ -1,46 +1,276 @@
 using Content.Shared.Interaction;
+using Content.Shared.Interaction.Events;
 using JetBrains.Annotations;
 using Robust.Server.GameObjects;
 using Robust.Shared.Player;
+using Robust.Shared.Utility;
+using static Content.Shared.Atmos.Components.SharedGasAnalyzerComponent;
 
 namespace Content.Server.Atmos.EntitySystems
 {
     [UsedImplicitly]
     public sealed class GasAnalyzerSystem : EntitySystem
     {
         [Dependency] private readonly PopupSystem _popup = default!;
+        [Dependency] private readonly AtmosphereSystem _atmo = default!;
+        [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+        [Dependency] private readonly UserInterfaceSystem _userInterface = default!;
+
+        private HashSet<EntityUid> _activeAnalyzers = new();
+        private RemQueue<EntityUid> _invalidAnalyzers = new();
 
         public override void Initialize()
         {
             base.Initialize();
 
             SubscribeLocalEvent<GasAnalyzerComponent, AfterInteractEvent>(OnAfterInteract);
+            SubscribeLocalEvent<GasAnalyzerComponent, GasAnalyzerRefreshMessage>(OnRefreshMessage);
+            SubscribeLocalEvent<GasAnalyzerComponent, GasAnalyzerDisableMessage>(OnDisabledMessage);
+            SubscribeLocalEvent<GasAnalyzerComponent, ComponentShutdown>(OnComponentShutdown);
+            SubscribeLocalEvent<GasAnalyzerComponent, DroppedEvent>(OnDropped);
+            SubscribeLocalEvent<GasAnalyzerComponent, UseInHandEvent>(OnUseInHand);
         }
 
         public override void Update(float frameTime)
         {
-            foreach (var analyzer in EntityManager.EntityQuery<GasAnalyzerComponent>(true))
+            if (!_activeAnalyzers.Any())
+                return;
+
+            foreach (var analyzer in _activeAnalyzers)
             {
-                analyzer.Update(frameTime);
+                if (_invalidAnalyzers.List != null && _invalidAnalyzers.List.Contains(analyzer))
+                    continue;
+
+                if(!UpdateAnalyzer(analyzer))
+                    _invalidAnalyzers.Add(analyzer);
             }
+
+            foreach (var invalidAnalyzer in _invalidAnalyzers)
+            {
+                _activeAnalyzers.Remove(invalidAnalyzer);
+            }
+
+            _invalidAnalyzers.List?.Clear();
         }
 
+        /// <summary>
+        /// Activates the analyzer when used in the world, scanning either the target entity or the tile clicked
+        /// </summary>
         private void OnAfterInteract(EntityUid uid, GasAnalyzerComponent component, AfterInteractEvent args)
         {
             if (!args.CanReach)
             {
                 _popup.PopupEntity(Loc.GetString("gas-analyzer-component-player-cannot-reach-message"), args.User, Filter.Entities(args.User));
                 return;
             }
+            ActivateAnalyzer(uid, component, args.User, args.Target);
+            OpenUserInterface(args.User, component);
```

Comment:
is there any reason activating the analyzer is separate from opening the UI? useinhand activates the analyzer but doesn't open the UI, which seems like incorrect behavior

Reply:
UseInHand opens the UI already due to having ActivatableUIComponent

Reply:
ah okay